pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go

```diff
@@ -67,48 +70,82 @@ func (ba *BalancedAllocation) ScoreExtensions() framework.ScoreExtensions {
-	memoryFraction := fractionOfCapacity(requested[v1.ResourceMemory], allocable[v1.ResourceMemory])
-	// This to find a node which has most balanced CPU, memory and volume usage.
-	if cpuFraction >= 1 || memoryFraction >= 1 {
-		// if requested >= capacity, the corresponding host should never be preferred.
-		return 0
+	resourceToFraction := make(map[v1.ResourceName]float64)
+	var totalFraction float64
+	for name, value := range requested {
+		if _, ok := allocable[name]; ok {
+			fraction := fractionOfCapacity(value, allocable[name])
+			if fraction >= 1 {
+				// if requested >= capacity, the corresponding host should never be preferred.
+				return 0
+			}
+			totalFraction = totalFraction + fraction
+			resourceToFraction[name] = fraction
+		} else {
+			klog.InfoS("requested resource not found for node score calculation", "resource", name)
+			return 0
+		}
 	}
 
 	if includeVolumes && utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes) && allocatableVolumes > 0 {
-		volumeFraction := float64(requestedVolumes) / float64(allocatableVolumes)
+		volumeFraction := fractionOfCapacity(int64(requestedVolumes), int64(allocatableVolumes))
 		if volumeFraction >= 1 {
 			// if requested >= capacity, the corresponding host should never be preferred.
 			return 0
 		}
-		// Compute variance for all the three fractions.
-		mean := (cpuFraction + memoryFraction + volumeFraction) / float64(3)
-		variance := float64((((cpuFraction - mean) * (cpuFraction - mean)) + ((memoryFraction - mean) * (memoryFraction - mean)) + ((volumeFraction - mean) * (volumeFraction - mean))) / float64(3))
-		// Since the variance is between positive fractions, it will be positive fraction. 1-variance lets the
-		// score to be higher for node which has least variance and multiplying it with `MaxNodeScore` provides the scaling
-		// factor needed.
-		return int64((1 - variance) * float64(framework.MaxNodeScore))
+		resourceToFraction["volume"] = volumeFraction
+		totalFraction = totalFraction + volumeFraction
 	}
 
+	// For backward compatibility, if there is only cpu and memory, we will use the legacy formula to calculate the score.
+	cpuFraction, cpuExist := resourceToFraction[v1.ResourceCPU]
+	memFraction, memExist := resourceToFraction[v1.ResourceMemory]
+
 	// Upper and lower boundary of difference between cpuFraction and memoryFraction are -1 and 1
 	// respectively. Multiplying the absolute value of the difference by `MaxNodeScore` scales the value to
 	// 0-MaxNodeScore with 0 representing well balanced allocation and `MaxNodeScore` poorly balanced. Subtracting it from
 	// `MaxNodeScore` leads to the score which also scales from 0 to `MaxNodeScore` while `MaxNodeScore` representing well balanced.
-	diff := math.Abs(cpuFraction - memoryFraction)
-	return int64((1 - diff) * float64(framework.MaxNodeScore))
+	if (len(resourceToFraction) == 2) && cpuExist && memExist {
```

Comment:
The new formula is already general enough to handle this case, isn't it?

Reply:
pls see my reply above.